cost_calculator/instance_selector.py

```diff
@@ -176,12 +194,69 @@ def get_cheapest_instance(self, cpu_request, memory_request, gpu_spec):
         lowest_price = 100000000.0
         for inst in matches:
             price, is_t_unlimited = self.price_for_cpu_spec(cpu_request, inst)
-            if price > 0.0 and price < lowest_price:
+            if 0.0 < price < lowest_price:
                 lowest_price = price
                 cheapest_instance = inst['instanceType']
                 if is_t_unlimited:
                     cheapest_instance += ' (unlimited)'
-        return cheapest_instance, lowest_price
+        lowest_spot_price = lowest_price
+        lowest_spot_price = min(lowest_spot_price, self.get_spot_price(cheapest_instance))
+        return cheapest_instance, lowest_price, lowest_spot_price
+
+
+class PriceGetter:
+    provider_keys_map = {
+        "azure": "azure",
+        "aws": "amazon",
+        "gce": "google"
+    }
+
+    def __init__(self, provider, redis_client):
+        self.provider = provider
+        self.redis_client = redis_client
+        self.key_pattern = "/banzaicloud.com/cloudinfo/providers/{provider}/regions/{region}/prices/{instance_type}"
+
+    def _get_azure_region_key(self, region):
+        return region.replace(" ", "").lower()
+
+    def _get_key(self, instance_type, region):
+        provider = self.provider_keys_map[self.provider]
+        if provider == "azure":
+            region = self._get_azure_region_key(region)
+        return self.key_pattern.format(provider=provider, region=region, instance_type=instance_type)
+
+    def _convert_entry_to_dict(self, data):
+        prices_str = data.decode('utf-8')
+        if "null" in prices_str:
+            prices_str = prices_str.replace("null", "{}")
+        print(f"got raw data: {prices_str}")
+        try:
+            prices = ast.literal_eval(prices_str)
+        except ValueError:
+            raise ValueError(f"cannot convert {prices_str} to dict")
+        return prices
+
+    def _get_lowest_spot_price(self, prices):
+        spot_prices = prices.get("spotPrice")
+        if spot_prices is None or len(spot_prices.values()) == 0:
+            # no spotPrices found, get on-demand price
+            return prices['onDemandPrice']
+        return min(spot_prices.values())
+
+    def _get_data_for_instance(self, instance_type, region):
+        key = self._get_key(region=region, instance_type=instance_type)
+        data = self.redis_client.get(key)
+        prices = self._convert_entry_to_dict(data)
+        return prices
+
+    def get_spot_price(self, instance_type, region):
+        prices = self._get_data_for_instance(instance_type, region)
+        spot_price = self._get_lowest_spot_price(prices)
+        return spot_price
```

Comment:
Does this return a time series of spot prices over time?

Reply:
No, it returns current spot price for given instance. Probably in kip we would like to tweak scraping behavior to have more up to date prices.